File: backend/app/schemas/trust.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator


class KnowledgeChunkSeed(BaseModel):
    code: str = Field(min_length=3, max_length=120)
    page_ref: str | None = Field(default=None, max_length=80)
    paragraph_ref: str = Field(min_length=1, max_length=160)
    content: str = Field(min_length=20, max_length=4000)
    keywords: list[str] = Field(min_length=1, max_length=30)


class PolicyDocumentSeed(BaseModel):
    code: str = Field(min_length=3, max_length=80)
    title: str = Field(min_length=3, max_length=300)
    issuing_authority: str = Field(min_length=2, max_length=200)
    category: str = Field(min_length=3, max_length=80)
    document_version: str = Field(min_length=2, max_length=64)
    publication_date: date
    effective_date: date
    expiry_date: date | None = None
    source_uri: str = Field(min_length=8, max_length=500)
    source_type: str = Field(min_length=3, max_length=40)
    applicable_audiences: list[str] = Field(min_length=1)
    applicable_regions: list[str] = Field(min_length=1)
    last_verified_date: date
    controlled_snapshot: bool
    metadata: dict[str, Any] = Field(default_factory=dict)
    chunks: list[KnowledgeChunkSeed] = Field(min_length=1)

    @field_validator("expiry_date")
    @classmethod
    def validate_expiry(cls, value: date | None, info: Any) -> date | None:
        effective = info.data.get("effective_date")
        if value is not None and effective is not None and value < effective:
            raise ValueError("expiry_date must not precede effective_date")
        return value


class ControlledKnowledgeDataset(BaseModel):
    dataset_version: str = Field(min_length=3, max_length=64)
    updated_at: date
    vectorizer_version: str = Field(min_length=3, max_length=64)
    source_summary: str = Field(min_length=10, max_length=800)
    documents: list[PolicyDocumentSeed] = Field(min_length=6)


class KnowledgeCatalogDocument(BaseModel):
    code: str
    title: str
    issuing_authority: str
    category: str
    document_version: str
    publication_date: date
    effective_date: date
    expiry_date: date | None
    applicable_audiences: list[str]
    applicable_regions: list[str]
    source_type: str
    source_uri: str
    last_verified_date: date | None
    content_hash: str
    chunk_count: int
    status: Literal["active", "expired", "not_yet_effective"]


class KnowledgeCatalogResponse(BaseModel):
    dataset_version: str
    updated_at: date
    retrieval_version: str
    vectorizer_version: str
    source_summary: str
    document_count: int
    active_document_count: int
    chunk_count: int
    quarantined_chunk_count: int
    categories: list[str]
    documents: list[KnowledgeCatalogDocument]
    runtime_network_required: Literal[False] = False


class KnowledgeSearchRequest(BaseModel):
    query: str = Field(min_length=2, max_length=500)
    as_of_date: date | None = None
    categories: list[str] = Field(default_factory=list, max_length=12)
    audiences: list[str] = Field(default_factory=list, max_length=12)
    regions: list[str] = Field(default_factory=list, max_length=12)
    limit: int = Field(default=5, ge=1, le=10)


class KnowledgeCitation(BaseModel):
    citation_id: str
    chunk_id: str
    chunk_code: str
    document_code: str
    title: str
    issuing_authority: str
    category: str
    source_uri: str
    source_type: str
    publication_date: date
    effective_date: date
    expiry_date: date | None
    last_verified_date: date | None
    applicable_audiences: list[str]
    applicable_regions: list[str]
    page_ref: str | None
    paragraph_ref: str
    document_version: str
    content_hash: str


class KnowledgeMatch(BaseModel):
    chunk_id: str
    chunk_code: str
    excerpt: str
    keyword_score: Decimal
    vector_score: Decimal
    metadata_score: Decimal
    combined_score: Decimal
    citation_id: str


class KnowledgeClaim(BaseModel):
    text: str
    citation_ids: list[str] = Field(min_length=1)


class KnowledgeSearchResponse(BaseModel):
    query: str
    as_of_date: date
    answer: str
    claims: list[KnowledgeClaim]
    matches: list[KnowledgeMatch]
    citations: list[KnowledgeCitation]
    insufficient_information: bool
    filtered_expired_count: int
    filtered_not_yet_effective_count: int
    filtered_quarantined_count: int
    retrieval_version: str
    vectorizer_version: str
    calculation_source: Literal["deterministic_hybrid_retrieval"] = "deterministic_hybrid_retrieval"
    limitations: list[str]


class GraphNode(BaseModel):
    id: str
    node_type: str
    label: str
    lane: Literal["family", "facts", "goals", "controls", "knowledge"]
    properties: dict[str, str | int | bool | None] = Field(default_factory=dict)
    source_entity_type: str
    source_entity_id: str | None = None
    calculation_source: str


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relation: str
    label: str
    evidence: list[str]


class GraphInference(BaseModel):
    code: str
    title: str
    conclusion: str
    evidence_node_ids: list[str] = Field(min_length=1)
    rule: str
    human_review_required: bool = False


class HouseholdGraphResponse(BaseModel):
    graph_id: str
    graph_version: str
    title: str
    subtitle: str
    as_of_date: date
    synthetic: bool
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    inferences: list[GraphInference]
    node_type_coverage: list[str]
    calculation_source: Literal["deterministic_relational_graph_service"] = (
        "deterministic_relational_graph_service"
    )
    limitations: list[str]


class IntakeDraftRequest(BaseModel):
    text: str = Field(min_length=2, max_length=2000)
    household_id: str | None = None


class ExtractedDraftField(BaseModel):
    code: str
    label: str
    value: str
    value_type: Literal["money", "count", "stage", "relationship", "text"]
    unit: str | None = None
    evidence: str
    confidence: Decimal
    confirmed: bool


class MissingDraftField(BaseModel):
    code: str
    label: str
    reason: str
    required_for: list[str]
    priority: int = Field(default=3, ge=1, le=5)
    follow_up_question: str = "请补充这一信息。"


class IntakeDraftResponse(BaseModel):
    draft_id: str
    household_id: str | None
    status: Literal["pending_confirmation", "partially_confirmed", "confirmed"]
    parser_version: str
    source_text_hash: str
    redacted_preview: str
    extracted_fields: list[ExtractedDraftField]
    missing_fields: list[MissingDraftField]
    confirmed_values: dict[str, str]
    contains_untrusted_instruction: bool
    confirmation_required: bool
    created_at: datetime
    confirmed_at: datetime | None
    boundary_note: str


class ConfirmIntakeDraftRequest(BaseModel):
    confirmed_values: dict[str, str] = Field(min_length=1, max_length=30)


class AgentSpecOut(BaseModel):
    code: str
    name: str
    purpose: str
    sequence: int
    input_schema_name: str
    input_json_schema: dict[str, Any]
    output_schema_name: str
    output_json_schema: dict[str, Any]
    allowed_tools: list[str] = Field(min_length=1)
    prohibited_actions: list[str] = Field(min_length=1)
    timeout_seconds: int
    failure_fallback: str
    audit_event: str


class AgentCatalogResponse(BaseModel):
    orchestrator_version: str
    state_machine: list[str]
    agent_count: Literal[9] = 9
    agents: list[AgentSpecOut]
    hard_gates: list[str]


class OrchestrationRequest(BaseModel):
    request_kind: Literal["trusted_plan_explanation", "policy_and_plan_review"] = (
        "trusted_plan_explanation"
    )
    policy_query: str = Field(
        default="个人养老金与家庭适配需要核对什么?", min_length=2, max_length=500
    )
    analysis_date: date | None = None


class ToolCallEvidence(BaseModel):
    tool: str
    status: Literal["completed", "degraded", "blocked"]
    input_reference: str
    output_reference: str
    calculation_source: str


class AgentStepOut(BaseModel):
    step_id: str
    agent_code: str
    agent_name: str
    sequence: int
    status: Literal["pending", "running", "completed", "degraded", "blocked"]
    input_schema_name: str
    output_schema_name: str
    tool_calls: list[ToolCallEvidence]
    structured_output: dict[str, Any]
    citations: list[str]
    prohibitions_checked: list[str]
    timeout_seconds: int
    failure_code: str | None
    degraded: bool
    started_at: datetime
    completed_at: datetime | None


class NumericLedgerEntry(BaseModel):
    code: str
    value: str
    unit: str
    source_tool: str
    source_path: str
    value_hash: str


class GovernanceIssue(BaseModel):
    code: str
    severity: Literal["info", "warning", "block"]
    message: str
    claim_index: int | None = None
    evidence: list[str] = Field(default_factory=list)


class OrchestrationResponse(BaseModel):
    run_id: str
    household_id: str
    request_kind: str
    status: Literal["running", "completed", "degraded", "blocked"]
    current_state: str
    orchestrator_version: str
    provider_mode: str
    steps: list[AgentStepOut]
    numeric_ledger: list[NumericLedgerEntry]
    citation_chunk_ids: list[str]
    structured_output: dict[str, Any]
    blocked_issues: list[GovernanceIssue]
    requires_human_review: bool
    degraded: bool
    started_at: datetime
    completed_at: datetime | None
    boundary_note: str


class GovernanceClaim(BaseModel):
    text: str = Field(min_length=1, max_length=1000)
    claim_type: Literal["numeric", "policy", "product", "general"]
    value: str | None = None
    tool_reference: str | None = None
    citation_chunk_ids: list[str] = Field(default_factory=list)
    product_code: str | None = None
    product_catalog_version: str | None = None


class GovernanceValidationRequest(BaseModel):
    claims: list[GovernanceClaim] = Field(min_length=1, max_length=30)
    numeric_ledger: list[NumericLedgerEntry] = Field(default_factory=list, max_length=100)
    as_of_date: date | None = None
    ordinary_household_path: bool = True


class GovernanceValidationResponse(BaseModel):
    passed: bool
    blocked: bool
    requires_human_review: bool
    issues: list[GovernanceIssue]
    validated_claim_count: int
    calculation_source: Literal["deterministic_governance_validator"] = (
        "deterministic_governance_validator"
    )

File: frontend/src/api/trust.ts
export interface KnowledgeCatalog {
  dataset_version: string;
  updated_at: string;
  retrieval_version: string;
  vectorizer_version: string;
  source_summary: string;
  document_count: number;
  active_document_count: number;
  chunk_count: number;
  quarantined_chunk_count: number;
  categories: string[];
  runtime_network_required: false;
}

export interface KnowledgeCitation {
  citation_id: string;
  chunk_id: string;
  chunk_code: string;
  document_code: string;
  title: string;
  issuing_authority: string;
  category: string;
  source_uri: string;
  source_type: string;
  publication_date: string;
  effective_date: string;
  expiry_date: string | null;
  last_verified_date: string | null;
  applicable_audiences: string[];
  applicable_regions: string[];
  page_ref: string | null;
  paragraph_ref: string;
  document_version: string;
  content_hash: string;
}

export interface KnowledgeSearch {
  query: string;
  as_of_date: string;
  answer: string;
  claims: Array<{ text: string; citation_ids: string[] }>;
  matches: Array<{
    chunk_id: string;
    chunk_code: string;
    excerpt: string;
    keyword_score: string;
    vector_score: string;
    metadata_score: string;
    combined_score: string;
    citation_id: string;
  }>;
  citations: KnowledgeCitation[];
  insufficient_information: boolean;
  filtered_expired_count: number;
  filtered_not_yet_effective_count: number;
  filtered_quarantined_count: number;
  retrieval_version: string;
  vectorizer_version: string;
  calculation_source: "deterministic_hybrid_retrieval";
  limitations: string[];
}

export type GraphLane = "family" | "facts" | "goals" | "controls" | "knowledge";

export interface HouseholdGraph {
  graph_id: string;
  graph_version: string;
  title: string;
  subtitle: string;
  as_of_date: string;
  synthetic: boolean;
  nodes: Array<{
    id: string;
    node_type: string;
    label: string;
    lane: GraphLane;
    properties: Record<string, string | number | boolean | null>;
    source_entity_type: string;
    source_entity_id: string | null;
    calculation_source: string;
  }>;
  edges: Array<{
    id: string;
    source: string;
    target: string;
    relation: string;
    label: string;
    evidence: string[];
  }>;
  inferences: Array<{
    code: string;
    title: string;
    conclusion: string;
    evidence_node_ids: string[];
    rule: string;
    human_review_required: boolean;
  }>;
  node_type_coverage: string[];
  calculation_source: "deterministic_relational_graph_service";
  limitations: string[];
}

export interface IntakeDraft {
  draft_id: string;
  household_id: string | null;
  status: "pending_confirmation" | "partially_confirmed" | "confirmed";
  parser_version: string;
  source_text_hash: string;
  redacted_preview: string;
  extracted_fields: Array<{
    code: string;
    label: string;
    value: string;
    value_type: "money" | "count" | "stage" | "relationship" | "text";
    unit: string | null;
    evidence: string;
    confidence: string;
    confirmed: boolean;
  }>;
  missing_fields: Array<{
    code: string;
    label: string;
    reason: string;
    required_for: string[];
    priority: number;
    follow_up_question: string;
  }>;
  confirmed_values: Record<string, string>;
  contains_untrusted_instruction: boolean;
  confirmation_required: boolean;
  created_at: string;
  confirmed_at: string | null;
  boundary_note: string;
}

export interface AgentCatalog {
  orchestrator_version: string;
  state_machine: string[];
  agent_count: 9;
  agents: Array<{
    code: string;
    name: string;
    purpose: string;
    sequence: number;
    input_schema_name: string;
    input_json_schema: Record<string, unknown>;
    output_schema_name: string;
    output_json_schema: Record<string, unknown>;
    allowed_tools: string[];
    prohibited_actions: string[];
    timeout_seconds: number;
    failure_fallback: string;
    audit_event: string;
  }>;
  hard_gates: string[];
}

export interface Orchestration {
  run_id: string;
  household_id: string;
  request_kind: string;
  status: "running" | "completed" | "degraded" | "blocked";
  current_state: string;
  orchestrator_version: string;
  provider_mode: string;
  steps: Array<{
    step_id: string;
    agent_code: string;
    agent_name: string;
    sequence: number;
    status: "pending" | "running" | "completed" | "degraded" | "blocked";
    input_schema_name: string;
    output_schema_name: string;
    tool_calls: Array<{
      tool: string;
      status: "completed" | "degraded" | "blocked";
      input_reference: string;
      output_reference: string;
      calculation_source: string;
    }>;
    structured_output: Record<string, unknown>;
    citations: string[];
    prohibitions_checked: string[];
    timeout_seconds: number;
    failure_code: string | null;
    degraded: boolean;
    started_at: string;
    completed_at: string | null;
  }>;
  numeric_ledger: Array<{
    code: string;
    value: string;
    unit: string;
    source_tool: string;
    source_path: string;
    value_hash: string;
  }>;
  citation_chunk_ids: string[];
  structured_output: Record<string, unknown>;
  blocked_issues: Array<{
    code: string;
    severity: "info" | "warning" | "block";
    message: string;
    claim_index: number | null;
    evidence: string[];
  }>;
  requires_human_review: boolean;
  degraded: boolean;
  started_at: string;
  completed_at: string | null;
  boundary_note: string;
}

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function readJson<T>(path: string, signal?: AbortSignal, init?: RequestInit): Promise<T> {
  const response = await fetch(apiBase + path, {
    ...init,
    signal,
    headers: { Accept: "application/json", ...init?.headers },
  });
  if (!response.ok) throw new Error(`API ${response.status}: ${path}`);
  return (await response.json()) as T;
}

export function fetchKnowledgeCatalog(signal?: AbortSignal): Promise<KnowledgeCatalog> {
  return readJson("/api/v1/trust/knowledge/catalog?as_of_date=2026-08-04", signal);
}

export function searchKnowledge(query: string, signal?: AbortSignal): Promise<KnowledgeSearch> {
  return readJson("/api/v1/trust/knowledge/search", signal, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query, as_of_date: "2026-08-04", limit: 5 }),
  });
}

export function fetchShanghaiGraph(signal?: AbortSignal): Promise<HouseholdGraph> {
  return readJson("/api/v1/trust/graphs/shanghai-demo?as_of_date=2026-08-04", signal);
}

export function fetchHouseholdGraph(
  householdId: string,
  signal?: AbortSignal,
): Promise<HouseholdGraph> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/trust-graph?analysis_date=2026-08-04`,
    signal,
  );
}

export function createIntakeDraft(householdId: string, text: string): Promise<IntakeDraft> {
  return readJson("/api/v1/trust/intake/drafts", undefined, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ household_id: householdId, text }),
  });
}

export function confirmIntakeDraft(
  draftId: string,
  confirmedValues: Record<string, string>,
): Promise<IntakeDraft> {
  return readJson(`/api/v1/trust/intake/drafts/${encodeURIComponent(draftId)}/confirm`, undefined, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ confirmed_values: confirmedValues }),
  });
}

export function fetchAgentCatalog(signal?: AbortSignal): Promise<AgentCatalog> {
  return readJson("/api/v1/trust/agents/catalog", signal);
}

export function fetchLatestOrchestration(
  householdId: string,
  signal?: AbortSignal,
): Promise<Orchestration | null> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/trust-orchestrations/latest`,
    signal,
  );
}

export function runTrustOrchestration(householdId: string): Promise<Orchestration> {
  return readJson(
    `/api/v1/households/${encodeURIComponent(householdId)}/trust-orchestrations`,
    undefined,
    {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        request_kind: "trusted_plan_explanation",
        policy_query: "个人养老金每年缴费限额和家庭适配需要核对什么?",
        analysis_date: "2026-08-04",
      }),
    },
  );
}
